src/InodeCache.cpp

```diff
@@ -0,0 +1,488 @@
+#  include <errno.h>
+#  include <fcntl.h>
+#  include <libgen.h>
+#  include <stdio.h>
+#  include <stdlib.h>
+#  include <string.h>
+#  include <sys/mman.h>
+#  include <sys/stat.h>
+#  include <sys/types.h>
+#  include <time.h>
+#  include <unistd.h>
+
+namespace {
+
+// The inode cache resides on a file that is mapped into shared memory by
+// running processes. It is implemented as a two level structure, where the
+// top level is a hash table consisting of buckets. Each bucket contains entries
+// that are sorted in LRU order. Entries map from keys representing files to
+// cached hash results.
+//
+// Concurrent access is guarded by a mutex in each bucket.
+//
+// Current cache size is fixed and the given constants are considered large
+// enough for most projects. The size could be made configurable if there is a
+// demand for it.
+const uint32_t k_version = 1;
+
+// Increment version number if constants affecting storage size are changed.
+const uint32_t k_num_buckets = 32 * 1024;
+const uint32_t k_num_entries = 4;
+
+static_assert(sizeof(digest::bytes) == 20,
+              "Increment version number if size of digest is changed.");
+
+static const std::string version_suffix =
+  std::string(".v") + std::to_string(k_version);
+
+// Return path to /run/user/{euid} if it exists since it is likely to be on
+// tmpfs, otherwise fall back on /tmp.
+std::string
+get_tmpdir()
+{
+  std::string tmpdir = "/run/user/" + std::to_string(geteuid());
+  if (!Stat::stat(tmpdir).is_directory())
+    tmpdir = "/tmp";
+  return tmpdir;
+}
+
+const std::string&
+get_default_filename()
+{
+  static const std::string filename =
+    get_tmpdir() + "/ccache_inode_cache" + version_suffix;
+  return filename;
+}
+
+std::string
+get_file_from_config(const Config& config)
+{
+  return config.inode_cache_file().empty()
+           ? get_default_filename()
+           : config.inode_cache_file() + version_suffix;
+}
+
+} // namespace
+
+struct InodeCache::Key
+{
+  dev_t st_dev;
+  ino_t st_ino;
+  mode_t st_mode;
+#  ifdef HAVE_STRUCT_STAT_ST_MTIM
+  timespec st_mtim;
+#  else
+  time_t st_mtim;
+#  endif
+#  ifdef HAVE_STRUCT_STAT_ST_CTIM
+  timespec st_ctim; // Included for sanity checking.
+#  else
+  time_t st_ctim; // Included for sanity checking.
+#  endif
+  off_t st_size; // Included for sanity checking.
+  bool sloppy_time_macros;
+};
+
+struct InodeCache::Entry
+{
+  digest key_digest;  // Hashed key
+  digest file_digest; // Cached file hash
+  int return_value;   // Cached return value
+};
+
+struct InodeCache::Bucket
+{
+  pthread_mutex_t mt;
+  int32_t hits;
+  int32_t misses;
+  Entry entries[k_num_entries];
+};
+
+struct InodeCache::SharedRegion
+{
+  uint32_t version;
+  std::atomic<int64_t> errors;
+  Bucket buckets[k_num_buckets];
+};
+
+bool
+InodeCache::mmap_file(const std::string& inode_cache_file)
+{
+  if (m_sr) {
+    munmap(m_sr, sizeof(SharedRegion));
+    m_sr = nullptr;
+  }
+  int fd = open(inode_cache_file.c_str(), O_RDWR);
```

Comment:
Need to use `O_RDWR | O_CREAT | O_EXCL, 0666` to avoid symlink attacks on files in `/tmp`.

Reply:
I changed it to work in this way:
1. Add uid to the filename if /tmp is used as default location to avoid name conflicts on systems with more than one user.
2. A new files is created using create_temp_fd, which uses mkstemp.
3. chmod the new file to 0600 not allowing others to access it.
4. The file is renamed using linkat(), which fails if a file or link with the target name already exists.
5. Existing files are opened using O_NOFOLLOW that fails it the target is a link.
6. Ownership is checked on the opened file descriptor and it is rejected if owned by someone else.

Reply:
Thanks. However, I'm wondering if it wouldn't be better to move it back to the cache directory again, the reason being @pavsa's [multi-user use case](https://github.com/ccache/ccache/issues/377#issuecomment-483785526). What do you think about that use case? (The fact that it doesn't work by default if the cache directory is on NFS is not an important problem, I'd say.)

Reply:
Could move it back of course. But I don't think performance is a good reason for sharing a single cache between users.

If we look at the numbers a large project will contain in the order of 10⁵ source files that need to be hashed. Without the cache we will hash the same files hundreds of times, ending up in the order of 10⁷-10⁸ calls to hash_source_code_file(). Since the cache will eliminate more than 99% of the calls, the remaining time spent in hashing is not be a big deal. We cannot save much by sharing the same cache between users because hashing is not a hot spot any longer.

But if you want me to change it back to the old location I will do that.

Reply:
Sound reasoning. Then the only advantage I see with having it in the cache directory is that it requires less complex code.

Reply:
Moved it back now and then we can reconsider the decision again if it ever becomes a problem.